nornir/core/inventory.py

```diff
@@ -230,43 +232,108 @@ def nos(self):
-            r = self.nornir.filter(name=self.name).run(conn_task, num_workers=1)
-            if r[self.name].exception:
-                raise r[self.name].exception
+            self.open_connection(
+                connection,
+                self.host,
+                self.username,
+                self.password,
+                self.ssh_port,
+                self.network_api_port,
+                self.os,
+                self.nos,
+                self.get(f"{connection}_options", {}),
+                config,
+            )
+        return self.connections[connection]
+
+    def open_connection(
+        self,
+        connection: str,
+        hostname: Optional[str] = None,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        ssh_port: Optional[int] = None,
+        network_api_port: Optional[int] = None,
+        operating_system: Optional[str] = None,
+        nos: Optional[str] = None,
+        connection_options: Optional[Dict[str, Any]] = None,
+        configuration: Optional[Config] = None,
+        default_to_host_attributes: bool = True,
+    ) -> None:
+        """
+        Open a new connection.
+
+        If ``default_to_host_attributes`` is set to ``True`` arguments will default to host
+        attributes if not specified.
 
+        Raises:
+            AttributeError: if it's unknown how to establish a connection for the given type
+
+        Returns:
+            An already established connection
+        """
+        self.connections[connection] = self.nornir.get_connection_type(connection)()
+        if default_to_host_attributes:
+            self.connections[connection].open(
+                hostname if hostname is not None else self.host,
+                username if username is not None else self.username,
+                password if password is not None else self.password,
+                ssh_port if ssh_port is not None else self.ssh_port,
+                network_api_port
+                if network_api_port is not None
+                else self.network_api_port,
+                operating_system if operating_system is not None else self.os,
+                nos if nos is not None else self.nos,
+                connection_options
+                if connection_options is not None
+                else self.get(f"{connection}_options"),
+                configuration if configuration is not None else self.nornir.config,
+            )
+        else:
+            self.connections[connection].open(
+                hostname,
+                username,
+                password,
+                ssh_port,
+                network_api_port,
+                operating_system,
+                nos,
+                connection_options,
+                configuration,
+            )
         return self.connections[connection]
 
+    def close_connection(self, connection: str) -> None:
+        """ Close the connection"""
+        self.connections[connection].close()
```

Comment:
When this is done, shouldn't it remove the dictionary entry from connections?

It surprised that after I did this:

```
task.host.close_connection("netmiko")
```

That I would then see:

```
(Pdb) task.host.connections
{'netmiko': <nornir.plugins.connections.netmiko.Netmiko object at 0xb640b6bc>}
(Pdb) task.host.connections['netmiko']
<nornir.plugins.connections.netmiko.Netmiko object at 0xb640b6bc>
(Pdb) task.host.connections['netmiko'].connection
<netmiko.cisco.cisco_ios.CiscoIosSSH object at 0xb640f12c>
(Pdb) task.host.connections['netmiko'].connection.remote_conn
(Pdb) p task.host.connections['netmiko'].connection.remote_conn
None
```

i.e. it was still in the Nornir connections dictionary even though there was no active connection.

Reply:
Related to this should `open_connection()` do any checking with respect to the connection already being in the `connections` table?

open_connection() will just create a new object regardless of whether an entry already exists in the connections table or not:

```
        self.connections[connection] = self.nornir.get_connection_type(connection)()
        if default_to_host_attributes:
```

Reply:
I agree with Kirk, I'd expect something like this:
```
    def close_connection(self, connection: str) -> None:
        self.connections.pop(connection).close()
```

Reply:
I disagree about `open_connection` though. IMO, `open_connection` should literally open a new connection. Plugins should use `get_connection`, which gets an existing connection if exists, or opens a new one.

Reply:
Re `close_connection`, yeah, we should remove it from available connections. Re `open_connection`, we could raise an exception if the connection is already open.